extensions/amp-analytics/0.1/session-manager.js

```diff
@@ -0,0 +1,252 @@
+   * @param {string|undefined} type
+   * @param {SESSION_VALUES} value
+   * @return {!Promise<number|undefined>}
+   */
+  getSessionValue(type, value) {
+    return this.get(type).then((session) => {
+      return session?.[value];
+    });
+  }
+
+  /**
+   * Get the session for the vendor, checking if it exists or
+   * creating it if necessary.
+   * @param {string|undefined} type
+   * @return {!Promise<SessionInfoDef|null>}
+   */
+  get(type) {
+    if (!type) {
+      user().error(TAG, 'Sessions can only be accessed with a vendor type.');
+      return Promise.resolve(null);
+    }
+
+    if (
+      hasOwn(this.sessions_, type) &&
+      !this.isSessionExpired_(this.sessions_[type])
+    ) {
+      this.setSession_(type, this.sessions_[type]);
+      this.sessions_[type].lastAccessTimestamp = Date.now();
+      return Promise.resolve(this.sessions_[type]);
+    }
+
+    return this.getSessionFromStorage_(type);
+  }
+
+  /**
+   * Get our session if it exists or creates it. Sets the session
+   * in localStorage to update the last access time.
+   * @param {string} type
+   * @return {!Promise<SessionInfoDef>}
+   */
+  getSessionFromStorage_(type) {
+    return this.storagePromise_
+      .then((storage) => {
+        const storageKey = getStorageKey(type);
+        // Gets the session if it has not expired
+        return storage.get(storageKey, SESSION_MAX_AGE_MILLIS);
+      })
+      .then((session) => {
+        // If no valid session in storage, create a new one
+        return !session
+          ? constructSessionInfo(generateSessionId(), Date.now())
+          : constructSessionFromStoredValue(session);
+      })
+      .then((session) => {
+        // Avoid multiple session creation race
+        if (
+          type in this.sessions_ &&
+          !this.isSessionExpired_(this.sessions_[type])
+        ) {
+          return this.sessions_[type];
+        }
+        this.setSession_(type, session);
+        session.lastAccessTimestamp = Date.now();
+        this.sessions_[type] = session;
+        return this.sessions_[type];
+      });
+  }
+
+  /**
+   * Set the session in localStorage, updating
+   * its last access time if it did not exist before.
+   * @param {string} type
+   * @param {SessionInfoDef} session
+   * @return {!Promise}
+   */
+  setSession_(type, session) {
+    return this.storagePromise_.then((storage) => {
+      const storageKey = getStorageKey(type);
+      const value = composeStorageSessionValue(session);
+      storage.setNonBoolean(storageKey, value);
+    });
+  }
+
+  /**
+   * Checks if a session has expired
+   * @param {SessionInfoDef} session
+   * @return {!Promise}
+   */
+  isSessionExpired_(session) {
+    return session.lastAccessTimestamp + SESSION_MAX_AGE_MILLIS < Date.now();
+  }
+}
+
+/**
+ * Return a pseudorandom low entropy value for session id.
+ * @return {number}
+ */
+function generateSessionId() {
+  return Math.floor(10000 * Math.random());
```

Comment:
This might create decimals, is that ok?

Reply:
`Math.floor()` generates an integer, so we should never get decimals right?

Reply:
Yep. `Math.round` might be better.